pkg/virt-launcher/virtwrap/network/network.go

```diff
@@ -35,103 +35,92 @@ import (
-// different context.
-// Phase1 is run by virt-handler and heavylifts most configuration steps. It
-// also creates the tap device that will be passed by name to virt-launcher,
-// thus allowing unprivileged libvirt to consume a pre-configured device.
-// Phase2 is run by virt-launcher in the pod context and completes steps left
-// out of virt-handler. The reason to have a separate phase for virt-launcher
-// and not just have all the work done by virt-handler is because there is no
-// ready solution for DHCP server startup in virt-handler context yet. This is
-// a temporary limitation and the split is expected to go once the final gap is
-// closed.
-// Moving all configuration steps into virt-handler will also allow to
-// downgrade privileges for virt-launcher, specifically, to remove NET_ADMIN
-// capability. Future patches should address that. See:
-// https://github.com/kubevirt/kubevirt/issues/3085
-type podNIC interface {
-	PlugPhase1(vmi *v1.VirtualMachineInstance, iface *v1.Interface, network *v1.Network, podInterfaceName string, pid int) error
-	PlugPhase2(vmi *v1.VirtualMachineInstance, iface *v1.Interface, network *v1.Network, domain *api.Domain, podInterfaceName string) error
+type VMNetworking struct {
+	vmi     *v1.VirtualMachineInstance
+	handler NetworkHandler
 }
 
-func getNetworksAndCniNetworks(vmi *v1.VirtualMachineInstance) (map[string]*v1.Network, map[string]int) {
-	networks := map[string]*v1.Network{}
-	cniNetworks := map[string]int{}
-	for _, network := range vmi.Spec.Networks {
-		networks[network.Name] = network.DeepCopy()
-		if networks[network.Name].Multus != nil && !networks[network.Name].Multus.Default {
-			// multus pod interfaces start from 1
-			cniNetworks[network.Name] = len(cniNetworks) + 1
-		}
+func NewVMNetworking(vmi *v1.VirtualMachineInstance) *VMNetworking {
+	return &VMNetworking{
+		vmi:     vmi,
+		handler: &NetworkUtilsHandler{},
 	}
-	return networks, cniNetworks
 }
 
-func invokePodNICFactory(networks map[string]*v1.Network, ifaceName string) (podNIC, error) {
-	network, ok := networks[ifaceName]
-	if !ok {
-		return nil, fmt.Errorf("failed to find a network %s", ifaceName)
+func (v VMNetworking) getNICs() ([]podNIC, error) {
+	nics := []podNIC{}
+
+	if len(v.vmi.Spec.Domain.Devices.Interfaces) == 0 {
+		return nics, nil
 	}
-	return podNICFactory(network)
-}
 
-func getPodInterfaceName(networks map[string]*v1.Network, cniNetworks map[string]int, ifaceName string) string {
-	if networks[ifaceName].Multus != nil && !networks[ifaceName].Multus.Default {
-		// multus pod interfaces named netX
-		return fmt.Sprintf("net%d", cniNetworks[ifaceName])
-	} else {
-		return primaryPodInterfaceName
+	idxMultus := 0
+	for i, network := range v.vmi.Spec.Networks {
+		if network.Pod == nil && network.Multus == nil {
+			return nil, fmt.Errorf("Network not implemented")
+		}
+		n := podNIC{handler: v.handler}
+		n.vmi = v.vmi
+		n.network = &n.vmi.Spec.Networks[i]
```

Comment:
Why don't you initialize all 3 attributes at once ? I personally find it simpler to understand.

Also 'n' could be improved; nic? iface ? podIface ? Literally, anything that doesn't shadow the `podNIC` struct name is better than `n`:)

Reply:
done